### ISSUE TITLE
Nomad 1.9.1: panic: runtime error: slice bounds out of range [:12] with capacity 0

### ISSUE THREAD

**Issue by HINT-SJ** [hashicorp/nomad#24379]
### Nomad version
```
Nomad v1.9.1
BuildDate 2024-10-21T09:00:50Z
Revision d9ec23f0c1035401e9df6c64d6ffb8bffc555a5e
```

### Operating system and Environment details
Amazon Linux 2023 (minimal)
AWS EC2 Gravitron (t4g.small)
> linux ip-#####.eu-central-1.compute.internal 6.1.112-122.189.amzn2023.aarch64 #1 SMP Tue Oct  8 17:01:34 UTC 2024 aarch64 aarch64 aarch64 GNU/Linux

### Issue

We observed two cases of Nomad clusters failing to start after a scheduled system patch window.
After installing system patches (through AWS SSM) and a followed system reboot (1 instance at a time, 5 instances acting as nomad server instances) we experienced the following panic upon service start on all 5 server instances:

```
nomad.service: Deactivated successfully.
Stopped nomad.service - Nomad Agent.
nomad.service: Consumed 13min 5.870s CPU time.
```

-- Instance Restart --

```
Started nomad.service - Nomad Agent.

==> WARNING: mTLS is not configured - Nomad is not secure without mTLS!
WARNING: keyring exists but -encrypt given, using keyring
==> Loaded configuration from /etc/nomad.d/nomad.hcl
==> Starting Nomad agent...
panic: runtime error: slice bounds out of range [:12] with capacity 0
goroutine 1 [running]:
github.com/hashicorp/go-kms-wrapping/v2/aead.(*Wrapper).Decrypt(0x40006cfa40, {0x40006cdc70?, 0x400091bb00?}, 0x400049e160, {0x0?, 0x20?, 0x21?})
        github.com/hashicorp/go-kms-wrapping/v2@v2.0.16/aead/aead.go:272 +0x1a0
github.com/hashicorp/nomad/nomad.(*Encrypter).loadKeyFromStore(0x4000685490, {0x400079c9c0?, 0x400079c9e0?})
        github.com/hashicorp/nomad/nomad/encrypter.go:809 +0x3a0
github.com/hashicorp/nomad/nomad.(*Encrypter).loadKeystore.func1({0x400079c9c0, 0x52}, {0x368c318?, 0x40009580c0?}, {0x0?, 0x0?})
        github.com/hashicorp/nomad/nomad/encrypter.go:172 +0x264
path/filepath.walk({0x400079c9c0, 0x52}, {0x368c318, 0x40009580c0}, 0x4000723100)
        path/filepath/path.go:345 +0xc8
path/filepath.walk({0x40004ad460, 0x1f}, {0x368c318, 0x4000133a40}, 0x4000723100)
        path/filepath/path.go:369 +0x1d0
path/filepath.Walk({0x40004ad460, 0x1f}, 0x4000723100)
        path/filepath/path.go:427 +0x6c
github.com/hashicorp/nomad/nomad.(*Encrypter).loadKeystore(0x4000685490)
        github.com/hashicorp/nomad/nomad/encrypter.go:145 +0xb4
github.com/hashicorp/nomad/nomad.NewEncrypter(0x4000427808, {0x40004ad460, 0x1f})
        github.com/hashicorp/nomad/nomad/encrypter.go:98 +0x19c
github.com/hashicorp/nomad/nomad.NewServer(0x400085d008, {0x3670460, 0x40001d7ef0}, 0x40009222b0, {0x368c828, 0x40001d7ee8})
        github.com/hashicorp/nomad/nomad/server.go:424 +0x9c8
github.com/hashicorp/nomad/command/agent.(*Agent).setupServer(0x4000766f00)
        github.com/hashicorp/nomad/command/agent/agent.go:971 +0x140
github.com/hashicorp/nomad/command/agent.NewAgent(0x400056fb08, {0x36b3608, 0x4000924c60}, {0x365c3c0, 0x40006924f8}, 0x40006cd680)
        github.com/hashicorp/nomad/command/agent/agent.go:157 +0x194
github.com/hashicorp/nomad/command/agent.(*Command).setupAgent(0x4000018580, 0x400056fb08, {0x36b3608, 0x4000924c60}, {0x365c3c0, 0x40006924f8}, 0x40006cd680)
        github.com/hashicorp/nomad/command/agent/command.go:639 +0x74
github.com/hashicorp/nomad/command/agent.(*Command).Run(0x4000018580, {0x400010c590, 0x1, 0x1})
        github.com/hashicorp/nomad/command/agent/command.go:856 +0x4c4
github.com/mitchellh/cli.(*CLI).Run(0x4000031900)
        github.com/mitchellh/cli@v1.1.5/cli.go:262 +0x488
main.Run({0x400010c580, 0x2, 0x2})
        github.com/hashicorp/nomad/main.go:112 +0x20c
main.main()
        github.com/hashicorp/nomad/main.go:82 +0x54

nomad.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
nomad.service: Failed with result 'exit-code'.
```

### Reproduction steps

4 of our 6 clusters went without any issues, just like before on version 1.8.1
So no real reproduction steps per se.

Our nomad.hcl:

```
data_dir = "/opt/nomad/data"
bind_addr = "0.0.0.0"

leave_on_interrupt = true
leave_on_terminate = true

ui {
  enabled =  true

  consul {
    ui_url = "https://consul.######/ui"
  }

  label {
    text   = "######"
  }
}

server {
  enabled = true
  bootstrap_expect = 5
  authoritative_region = "global"

  default_scheduler_config {
    scheduler_algorithm = "spread"
    preemption_config {
      batch_scheduler_enabled   = true
      system_scheduler_enabled  = true
      service_scheduler_enabled = true
    }
  }

  encrypt = "######"
}

http_api_response_headers {
  "Access-Control-Allow-Origin" = "*"
}

acl {
  enabled = true
}

telemetry {
  publish_allocation_metrics = true
  publish_node_metrics = true
  disable_hostname = true
  prometheus_metrics = true
}
```

#### Expected Result

No panic :)

#### Actual Result

> panic: runtime error: slice bounds out of range [:12] with capacity 0

:(

**Comment by blalor** [comment]
I hit the same issue, restarting a server to pick up new config. The binary was installed from the HC apt repo and I'm using the default systemd config. This is the journal output from when the server was stopped and then failed to start:

```
Nov 07 01:05:14 scribbler systemd[1]: Stopping Nomad...
Nov 07 01:05:14 scribbler nomad[1525905]: ==> Caught signal: interrupt
Nov 07 01:05:14 scribbler nomad[1525905]:     2024-11-07T01:05:14.455Z [INFO]  agent: requesting shutdown
Nov 07 01:05:14 scribbler nomad[1525905]:     2024-11-07T01:05:14.455Z [INFO]  nomad: shutting down server
Nov 07 01:05:14 scribbler nomad[1525905]:  agent: requesting shutdown
Nov 07 01:05:14 scribbler nomad[1525905]:  nomad: shutting down server
Nov 07 01:05:14 scribbler nomad[1525905]:     2024-11-07T01:05:14.810Z [WARN]  nomad: serf: Shutdown without a Leave
Nov 07 01:05:14 scribbler nomad[1525905]:  nomad: serf: Shutdown without a Leave
Nov 07 01:05:14 scribbler nomad[1525905]:     2024-11-07T01:05:14.810Z [INFO]  agent: shutdown complete
Nov 07 01:05:14 scribbler nomad[1525905]:  agent: shutdown complete
Nov 07 01:05:14 scribbler systemd[1]: nomad.service: Main process exited, code=exited, status=1/FAILURE
Nov 07 01:05:14 scribbler systemd[1]: nomad.service: Failed with result 'exit-code'.
Nov 07 01:05:14 scribbler systemd[1]: Stopped Nomad.
Nov 07 01:05:14 scribbler systemd[1]: nomad.service: Consumed 5min 49.251s CPU time.
Nov 07 01:05:14 scribbler systemd[1]: Starting Nomad...
Nov 07 01:05:14 scribbler nomad[2782173]: ==> WARNING: mTLS is not configured - Nomad is not secure without mTLS!
Nov 07 01:05:14 scribbler nomad[2782173]: ==> Config enable_syslog is `true` with log_level=INFO
Nov 07 01:05:14 scribbler nomad[2782173]: ==> Loaded configuration from /etc/nomad.d/00-main.hcl, /etc/nomad.d/10-server.hcl
Nov 07 01:05:14 scribbler nomad[2782173]: ==> Starting Nomad agent...
Nov 07 01:05:14 scribbler nomad[2782173]: panic: runtime error: slice bounds out of range [:12] with capacity 0
Nov 07 01:05:14 scribbler nomad[2782173]: goroutine 1 [running]:
Nov 07 01:05:14 scribbler nomad[2782173]: github.com/hashicorp/go-kms-wrapping/v2/aead.(*Wrapper).Decrypt(0xc000cac300, {0xc000ca1810?, 0xc000b05da0?}, 0xc000a0e4d0, {0x0?, 0x20?, 0x21?})
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/go-kms-wrapping/v2@v2.0.16/aead/aead.go:272 +0x199
Nov 07 01:05:14 scribbler nomad[2782173]: github.com/hashicorp/nomad/nomad.(*Encrypter).loadKeyFromStore(0xc0009e2e70, {0xc000502060?, 0xc00050207f?})
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/nomad/nomad/encrypter.go:809 +0x459
Nov 07 01:05:14 scribbler nomad[2782173]: github.com/hashicorp/nomad/nomad.(*Encrypter).loadKeystore.func1({0xc000502060, 0x51}, {0x3925278?, 0xc000c396c0?}, {0x0?, 0x0?})
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/nomad/nomad/encrypter.go:172 +0x25d
Nov 07 01:05:14 scribbler nomad[2782173]: path/filepath.walk({0xc000502060, 0x51}, {0x3925278, 0xc000c396c0}, 0xc000a21128)
Nov 07 01:05:14 scribbler nomad[2782173]:         path/filepath/path.go:345 +0x105
Nov 07 01:05:14 scribbler nomad[2782173]: path/filepath.walk({0xc0009fce60, 0x1e}, {0x3925278, 0xc000c395f0}, 0xc000a21128)
Nov 07 01:05:14 scribbler nomad[2782173]:         path/filepath/path.go:369 +0x254
Nov 07 01:05:14 scribbler nomad[2782173]: path/filepath.Walk({0xc0009fce60, 0x1e}, 0xc000a21128)
Nov 07 01:05:14 scribbler nomad[2782173]:         path/filepath/path.go:427 +0x66
Nov 07 01:05:14 scribbler nomad[2782173]: github.com/hashicorp/nomad/nomad.(*Encrypter).loadKeystore(0xc0009e2e70)
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/nomad/nomad/encrypter.go:145 +0xdf
Nov 07 01:05:14 scribbler nomad[2782173]: github.com/hashicorp/nomad/nomad.NewEncrypter(0xc0003fec08, {0xc0009fce60, 0x1e})
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/nomad/nomad/encrypter.go:98 +0x1a7
Nov 07 01:05:14 scribbler nomad[2782173]: github.com/hashicorp/nomad/nomad.NewServer(0xc0008de808, {0x39096e0, 0xc00009d200}, 0xc00099fb30, {0x3925788, 0xc00009d1f8})
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/nomad/nomad/server.go:424 +0xc50
Nov 07 01:05:14 scribbler nomad[2782173]: github.com/hashicorp/nomad/command/agent.(*Agent).setupServer(0xc0009ee700)
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/nomad/command/agent/agent.go:971 +0x17f
Nov 07 01:05:14 scribbler nomad[2782173]: github.com/hashicorp/nomad/command/agent.NewAgent(0xc000228248, {0x394c568, 0xc000bb2360}, {0x38f5800, 0xc000456048}, 0xc00076a140)
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/nomad/command/agent/agent.go:157 +0x1a7
Nov 07 01:05:14 scribbler nomad[2782173]: github.com/hashicorp/nomad/command/agent.(*Command).setupAgent(0xc000997200, 0xc000228248, {0x394c568, 0xc000bb2360}, {0x38f5800, 0xc000456048}, 0xc00076a140)
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/nomad/command/agent/command.go:639 +0xa5
Nov 07 01:05:14 scribbler nomad[2782173]: github.com/hashicorp/nomad/command/agent.(*Command).Run(0xc000997200, {0xc0000522e0, 0x2, 0x2})
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/nomad/command/agent/command.go:856 +0x585
Nov 07 01:05:14 scribbler nomad[2782173]: github.com/mitchellh/cli.(*CLI).Run(0xc000c23180)
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/mitchellh/cli@v1.1.5/cli.go:262 +0x55f
Nov 07 01:05:14 scribbler nomad[2782173]: main.Run({0xc0000522d0, 0x3, 0x3})
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/nomad/main.go:112 +0x219
Nov 07 01:05:14 scribbler nomad[2782173]: main.main()
Nov 07 01:05:14 scribbler nomad[2782173]:         github.com/hashicorp/nomad/main.go:82 +0x45
Nov 07 01:05:14 scribbler systemd[1]: nomad.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Nov 07 01:05:14 scribbler systemd[1]: nomad.service: Failed with result 'exit-code'.
Nov 07 01:05:14 scribbler systemd[1]: Failed to start Nomad.
Nov 07 01:05:16 scribbler systemd[1]: nomad.service: Scheduled restart job, restart counter is at 1.
Nov 07 01:05:16 scribbler systemd[1]: Stopped Nomad.
Nov 07 01:05:16 scribbler systemd[1]: Starting Nomad...
```

**Comment by jrasell** [comment]
Hi @blalor and @HINT-SJ, thanks for raising this issue.

I've been looking through the code path and have been able to reproduce the panic via a test but not running Nomad locally. In my test I am purposely "corrupting" the key wrapper I pass to the decryption process.

```go
func TestEncrypter_loadKeyFromStore_24379(t *testing.T) {
	ci.Parallel(t)

	srv := &Server{
		logger: testlog.HCLogger(t),
		config: &Config{},
	}

	tmpDir := t.TempDir()

	key, err := structs.NewUnwrappedRootKey(structs.EncryptionAlgorithmAES256GCM)
	must.NoError(t, err)

	encrypter, err := NewEncrypter(srv, tmpDir)
	must.NoError(t, err)

	wrappedKey, err := encrypter.encryptDEK(key, &structs.KEKProviderConfig{})
	diskWrapper := &structs.KeyEncryptionKeyWrapper{
		Meta:             key.Meta,
		KeyEncryptionKey: wrappedKey.KeyEncryptionKey,
		//EncryptedDataEncryptionKey: wrappedKey.WrappedDataEncryptionKey.Ciphertext,
		//EncryptedRSAKey:            wrappedKey.WrappedRSAKey.Ciphertext,
	}

	buf, err := json.Marshal(diskWrapper)
	must.NoError(t, err)

	path := filepath.Join(tmpDir, key.Meta.KeyID+".nks.json")
	err = os.WriteFile(path, buf, 0o600)
	must.NoError(t, err)

	encrypter.loadKeyFromStore(path)
}
``` 

You both mention this problem was reached after an upgrade, could you let me know what version you upgraded from please?

I will chat internally with the team about this and look into raising a potential fix, so that the servers do not panic when they come across this condition.

**Comment by HINT-SJ** [comment]
@jrasell we upgraded from 1.8.1 a week ago, the panic happened with 1.9.1 shutdown and starup again.

When some file is corrupted.. maybe there's some issue during shutdown (or release of a used file) of the process?

**Comment by jrasell** [comment]
@HINT-SJ thanks for the detail. I have just been doing some testing on 1.8.1 -> 1.9.1 upgrading and have not come across this problem yet. Did you upgrade to 1.8.1 or was this a new cluster? If you did upgrade, do you remember what version from? I can try testing this full upgrade path and see if that helps me discover it.

The file might not be exactly corrupted, but from my test reproduction, it seems it might be missing an item of content which seems quite odd. If you still have an impacted server it would be extremely useful if you could share some information about the files within the `server/keystore` directory. **Importantly here, I do not want to know and please do not share any real values from within the files, these are sensitive!** I am keen to understand what keys are in the files and if they have non-empty data. Please feel free to use the nomad-oss-debug at hashicorp.com address to send this.

Eg:
- Key "DEK" has data.
- Key "KEK: has no data / empty string.

**Comment by HINT-SJ** [comment]
@jrasell We upgraded our 1.8.1 clusters to 1.9.1 (went without any issue!) (clusters were created with version 0.7 or something ^^)

Unfortunately I don't have those files at hand anymore, but I'll see if I can reproduce it with one of our test clusters on 1 single instance

**Comment by blalor** [comment]
I'm _pretty sure_ I saw this same crash in an earlier version during my cluster upgrade, but I didn't save the stack trace; `go-kms-wrapping` looks familiar, however.

A couple of days ago I started upgrading a 3-server cluster running 1.2.6 (with some mixed client versions); the in-place upgrade path was 1.2.6 → 1.3.16 → 1.5.17 → 1.7.7 → 1.8.4.  Then I brought up 3 new servers running 1.9.1; they were happy members of the cluster and was in the process of removing the old (now 1.8.4) servers when I saw the crash.  The previous crash was *I think* when starting up one of the servers after upgrading from 1.7.7 to 1.8.4, but I don't recall exactly.  But I don't think this bug was introduced in 1.9.1.

But to be clear, this particular crash came after restarting one of the new 1.9.1 servers with just a configuration change (updating `server.server_join.retry_join` to contain the new list of servers, which triggered a restart, not a reload).

I'm able to reproduce the issue by just `systemctl restart nomad` on a server running 1.9.1. While it's crashing, I see a single file `/var/lib/nomad/server/keystore/XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX.aead.nks.json`with contents

```
{
  "Meta": {
    "KeyID": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
    "Algorithm": "aes256-gcm",
    "CreateTime": 1730822015041833700,
    "CreateIndex": 1577766,
    "ModifyIndex": 1577766,
    "State": "active",
    "PublishTime": 0
  },
  "ProviderID": "aead",
  "WrappedDEK": {
    "ciphertext": "…",
    "key_info": {
      "key_id": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
    }
  },
  "WrappedRSAKey": {},
  "KEK": "…"
}
```

I removed that file and the server was able to start again.
